### DISCUSSION TITLE
How to setup FastAPI to Heroku?

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#7864] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

Hey guys!

So I have been figured out how to code fastAPI and I am ready to deploy to heroku my script that I have worked with fastAPI however the problem is that when I do a request to the heroku it will just return:

```
<html>
  <head>
    <title>Internal Server Error</title>
  </head>
  <body>
    <h1><p>Internal Server Error</p></h1>
    
  </body>
</html>
```

and I am not quite able to see any logs where the problem is however I would say my problem might be that I am not sure if my **procfile** is correct because I haven't edit it at all and I am quite new at this and I am here to ask how I am able to run my fastapi script in heroku?

**Comment by wladneto** [comment]
Hello friend!

1- Add procfile into root project like this:

`web: uvicorn main:app --host=0.0.0.0 --port=${PORT:-5000}` 

2 - Add uvicorn (install) in your requirements.txt like this:

`uvicorb==0.9.0`

**Comment by windson** [comment]
You may want to go through this detailed video on [Deploy FastAPI on Heroku in just 6 minutes.](http://bit.ly/3aO1g35)

But gunicorn is used in this tutorial. In the Procfile, add the following command

`web: gunicorn -w 4 -k uvicorn.workers.UvicornWorker main:app`

You can also try @wladneto approach of specifying host and port in the gunicorn command and reply back if that worked for you.

**Comment by searchformyusername** [comment]
> 
> 
> Hello friend!
> 
> 1- Add procfile into root project like this:
> 
> `web: uvicorn main:app --host=0.0.0.0 --port=${PORT:-5000}`
> 
> 2 - Add uvicorn (install) in your requirements.txt like this:
> 
> `uvicorb==0.9.0`

it gives only 1 worker, I added `-w 4` in this which lead app to close. Any suggesstions?

**Comment by ycd** [comment]
@1UC1F3R616 try this procfile

```
web: gunicorn -w 4 -k uvicorn.workers.UvicornWorker main:app
```

or this

```
web: uvicorn main:app --workers 4
```

**Comment by searchformyusername** [comment]
Yep, I tried second and it worked. Thanks! @ycd

**Comment by searchformyusername** [comment]
> @1UC1F3R616 try this procfile
> 
> 
> 
> ```
> 
> web: gunicorn -w 4 -k uvicorn.workers.UvicornWorker main:app
> 
> ```
> 
> 
> 
> or this
> 
> 
> 
> ```
> 
> web: uvicorn main:app --workers 4
> 
> ```
> 
> 

First gives me error, what could be the reason? I have gunicron as requirements. In logs it shows worker crashed

**Comment by jarmos-san** [comment, marked as answer]
Hello everyone & whoever else stumbles across this thread through a Google Search!

**DISCLAIMER**: Heroku no longer supports a free tier anymore (see [this FAQ](https://help.heroku.com/RSBRUH58/removal-of-heroku-free-product-plans-faq)), hence I can longer experiment on the platform. As such, I cannot guaranteed the solution provided in the article linked below will work for you or not!

I had my fair share of trouble deploying to Heroku using GitHub Actions but I finally figured it out. So, I compiled every suggestion provided above; @wladneto's https://github.com/tiangolo/fastapi/issues/802#issuecomment-574429540 & @ycd's https://github.com/tiangolo/fastapi/issues/802#issuecomment-716034155 into an article.

I've also put together additional info which I found across other sources. So, feel free to use this article; [Using GitHub Actions to Deploy a FastAPI Project to Heroku](https://jarmos.netlify.app/posts/using-github-actions-to-deploy-a-fastapi-project-to-heroku/) as a starting point to deploy your project on Heroku.

**Comment by GTK-ARJUN** [comment]
> Hello friend!
> 
> 1- Add procfile into root project like this:
> 
> `web: uvicorn main:app --host=0.0.0.0 --port=${PORT:-5000}`
> 
> 2 - Add uvicorn (install) in your requirements.txt like this:
> 
> `uvicorb==0.9.0`

Please confirm that it's uvicorn and not uvicorb.

**Comment by Kludex** [comment]
Yes... It's uvicorn...
But 0.9.0 is too old, use 0.15.0.

**Comment by GTK-ARJUN** [comment]
Thank you so much as I was learning to deploy an app using uvicorn.

**Comment by Kludex** [comment]
No problem! Any doubts, feel free to ask! 

We're also on Discord in case you prefer!

**Comment by GTK-ARJUN** [comment]
Please add me or tell me how to join.
GTK_ARJUN_AGS#1771

**Comment by Kludex** [comment]
https://fastapi.tiangolo.com/help-fastapi/?h=discord#join-the-chat

**Comment by ArcLightSlavik** [comment]
@tiangolo @Kludex maybe close this since theres many articles out already?

**Comment by tiangolo** [comment]
Thanks for the discussion everyone! As the original poster is no longer on GitHub and it seems the solution was found, I'll close this issue now. ☕

**Comment by sergenti** [comment]
quick question, is `--worker-class=uvicorn.workers.UvicornWorker` async?

I was reading here https://docs.gunicorn.org/en/latest/design.html#async-workers that the default gunicorn worker does not support async tasks

**Reply by jarmos-san** [reply]
Yes, [`uvicorn`](https://www.uvicorn.org/) is general is async and from the looks of the code snippet you shared it looks like you're using `gunicorn` to spin up some async worker instances through `uvicorn`.
